aten/src/ATen/native/SobolEngineOps.cpp

```diff
@@ -0,0 +1,147 @@
+#include "ATen/ATen.h"
+#include "ATen/NativeFunctions.h"
+
+#include "ATen/native/SobolEngineOpsUtils.h"
+
+#include <vector>
+
+namespace at {
+namespace native {
+
+/// This is the core function to draw samples from a `SobolEngine` given
+/// its state variables (`sobolstate` and `quasi`). `dimension` can be
+/// inferred from `sobolstate`, but choosing to pass it explicitly to avoid
+/// an extra operation to obtain the size of the first dimension of
+/// `sobolstate`.
+std::tuple<Tensor, Tensor> _sobol_engine_draw(const Tensor& quasi, int64_t n, const Tensor& sobolstate,
+                                              int64_t dimension, int64_t num_generated) {
+  AT_CHECK(sobolstate.type().scalarType() == at::ScalarType::Long,
+           "sobolstate needs to be of type ", at::ScalarType::Long);
+  AT_CHECK(quasi.type().scalarType() == at::ScalarType::Long,
+           "quasi needs to be of type ", at::ScalarType::Long);
+
+  /// Performing one `unbind` operation and caching the result to prevent `n`
+  /// `select` operations.
+  std::vector<Tensor> sobolstate_unbind = at::native::unbind(sobolstate, 1);
+
+  /// Considering a vector of `n` Tensors to store the results in.
+  std::vector<Tensor> result;
+  Tensor wquasi = quasi.clone();
+
+  for (int64_t i = 0; i < n; ++i) {
+    int64_t l = rightmost_zero(num_generated);
+    result.emplace_back(wquasi.__ixor__(sobolstate_unbind[l]).clone());
```

Comment:
This needs to be optimized. I tried using a `copy_` instead but I kept getting an error about `copy_from` not being auto-differentiable.

Reply:
Hmm nothing here in this PR should need to be differentiable / require gradients at all. Is this creating a bunch of overhead by allowing for that?

Reply:
I don't know. It seemingly should not be the case, but I wonder if it is now. I disabled generating autograd functions for these functions, but I am puzzled.

Reply:
@soumith Is there any reason why autograd is complaining despite the fact that there are no gradients / tensors that require grad anywhere?

Reply:
@cpuhrsch any ideas what's going on here?